proposals/NNNN-rangeset-and-collection-operations.md

```diff
@@ -0,0 +1,616 @@
+let notTheMiddle: RangeSet = [0..<5, 10..<15]
+print(Array(numbers[notTheMiddle]))
+// Prints [1, 2, 3, 4, 5, 11, 12, 13, 14, 15]
+
+// ...or by using set operations
+let smallEvens = indicesOfEvens.intersection(
+    numbers.indices(where: { $0 < 10 }))
+print(Array(numbers[smallEvens]))
+// Prints [2, 4, 6, 8]
+```
+
+These are just a few examples; the complete proposal includes operations like inverting a range set and inserting and removing ranges, which are covered in the next section.
+
+## Detailed design
+
+`RangeSet` is generic over any `Comparable` type, and supports fast containment checks for ranges and individual values, as well as adding and removing ranges of that type. 
+
+```swift
+/// A set of ranges of any comparable value.
+public struct RangeSet<Bound: Comparable> {
+    /// Creates an empty range set.
+    public init() {}
+
+    /// Creates a range set containing the given range.
+    public init(_ range: Range<Bound>)
+    
+    /// Creates a range set containing the given ranges.
+    public init<S: Sequence>(_ ranges: S) where S.Element == Range<Bound>
+    
+    /// A Boolean value indicating whether the range set is empty.
+    public var isEmpty: Bool { get }
+    
+    /// Returns a Boolean value indicating whether the given value is
+    /// contained by the ranges in the range set.
+    ///
+    /// - Complexity: O(log *n*), where *n* is the number of ranges in the
+    ///   range set.
+    public func contains(_ value: Bound) -> Bool
+        
+    /// Returns a Boolean value indicating whether the given range is
+    /// contained by the ranges in the range set.
+    ///
+    /// This method returns `true` if and only if one of the ranges that 
+    /// comprises the range set fully contains `range`. This example shows
+    /// ranges that are and are not contained by a range set:
+    /// 
+    ///     let set: RangeSet = [0..<5, 10..<15]
+    ///     print(set.contains(1..<4))
+    ///     // Prints "true"
+    ///     print(set.contains(4..<7))
+    ///     // Prints "false"
+    ///     print(set.contains(4..<12))
+    ///     // Prints "false"
+    ///
+    /// - Complexity: O(log *n*), where *n* is the number of ranges in the
+    ///   range set.
+    public func contains(_ range: Range<Bound>) -> Bool
+    
+    /// Inserts the given range into the range set.
+    public mutating func insert(_ range: Range<Bound>)
+        
+    /// Removes the given range from the range set.
+    public mutating func remove(_ range: Range<Bound>)
+}
+```
+
+`RangeSet` conforms to `Equatable`, to `CustomStringConvertible`, and, when its `Bound` type conforms to `Hashable`, to `Hashable`. `RangeSet` also has `ExpressibleByArrayLiteral` conformance, using arrays of ranges as its literal type.
+
+#### `SetAlgebra`-like methods
+
+`RangeSet` implements a subset of the `SetAlgebra` protocol. Because of its efficient representation for contiguous ranges, there is no way to insert a single element into a `RangeSet` .
```

Comment:
If `RangeSet` is indeed a *set of ranges* as noted in the doc comment, then *elements* of the set have to be ranges. So isn't what you mean that there's no way to insert a range containing just one `Bound` value?